src/generic/GenericTypes.jl

```diff
@@ -798,6 +798,44 @@ end
 
 const PuiseuxSeriesElem{T} = Union{PuiseuxSeriesRingElem{T}, PuiseuxSeriesFieldElem{T}} where T <: RingElement
 
+###############################################################################
+#
+#   AbsMSeriesRing / AbsMSeries
+#
+###############################################################################
+
+mutable struct AbsMSeriesRing{T <: RingElement,
+           S <: AbstractAlgebra.MPolyRing{T}} <: AbstractAlgebra.MSeriesRing{T}
+   base_ring::Ring
+   poly_ring::S
+   prec_max::Vector{Int}
+   sym::Vector{Symbol}
+
+   function AbsMSeriesRing{T, S}(R::Ring, poly_ring::S, prec::Vector{Int},
+                            s::Vector{Symbol}, cached::Bool = true) where
+                          {T <: RingElement, S <: AbstractAlgebra.MPolyRing{T}}
+      return get_cached!(AbsMSeriesID, (R, prec, s), cached) do
+         new{T, S}(R, poly_ring, prec, s)
+      end::AbsMSeriesRing{T}
+   end
+end
+
+const AbsMSeriesID = CacheDictType{Tuple{Ring,
+                                          Vector{Int}, Vector{Symbol}}, Ring}()
+
+mutable struct AbsMSeries{T <: RingElement,
+        S <: AbstractAlgebra.MPolyElem{T}} <: AbstractAlgebra.AbsMSeriesElem{T}
+   poly::S
+   prec::Vector{Int}
+   parent::AbsMSeriesRing{T}
+
+   function AbsMSeries{T, S}(p::AbstractAlgebra.MPolyElem{T},
+                             prec::Vector{Int}) where
+                          {T <: RingElement, S <: AbstractAlgebra.MPolyElem{T}}
+      return new{T, S}(p, prec)
```

Comment:
Are there cases where one would want an `AbsMSeries` with uninitialized parent? Otherwise, it would seem better to initialize it in this inner constructor, which presumably would also allow you to have concrete types in the structs.

Reply:
There are no such cases. So I can simply pass the parent to the inner constructor.

Reply:
As for having concrete types in the structs, I don't think this is a good idea. We'll end up with a four parameter type on each of the parent and element types, and it's just a mess.

The types in the parent struct are not type stable anyway due to the base_ring just belonging to Ring, as is the case with every other parent struct.

I am going back to a one parameter type for each, sorry @thofma. This is just too messy otherwise.

Reply:
The concrete type for the poly is enough, like we do for everything else. Because this is what always will be accessed.

Reply:
OK, if this is too messy, then we should at least turn `f.poly` into `poly(f)` and sprinkle the type assertion into the `poly(f..)` series. Otherwise we end up with the same inference/compiler problems we had with the modules.

Reply:
The other option would be to have both the parent and element types parameterised by both the poly element and poly ring types and not by the type of the coefficient ring at all.

Let me know what you guys think.

Reply:
The problem of course is that whatever types we use, we need to implement the parent_type and elem_type functions. Moreover, whatever type the series parent object has, this type needs to appear (uncomputed) in the parent field of the series element type.

That basically means we need a three parameter type for each if we want it all to be type stable.

Reply:
@thofma There is already a poly(f). I can add a type assertion to it.

Reply:
Hmm, actually, there is no way to compute the poly type from the coefficient type.

Reply:
> not by the type of the coefficient ring at all.

Seems to work, no?
This leads to `AbsMSeriesRing{T, PR}` being parameterized by `T`, which can be computed from `poly_ring` and `PR`, the type of `poly_ring`. `R` can be removed, as it's `base_ring(poly_ring)` so is redundant (sorry if I'm off here, it's easy to get confused). 
`AbsMSeries` seems indeed to need 3 type parameters to have concrete field types.

Reply:
I guess we would need to define `mpoly_type(T)`, which would need to be implemented and rolled out, similar to `dense_matrix_type` etc.

Reply:
Ok, here is what I did to fix this. Both the parent and element AbsMSeries types are now parameterised by both the type of the coefficient elements and the poly ring. That makes AbsMSeries type stable.

The AbsMSeriesRing type accesses its poly_ring and base_ring fields only through functions which now have type assertions to help the compiler.

I think this is basically how all types are going to have to work that wrap another complex type which has its own coefficient type {mumbles something or other about Laurent series and Puiseux series}.